File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int i;
    int find(int x,vector<int>in){
        for(int i=0;i<in.size();i++){
            if(in[i]==x) return i;
        }
        return -1;
    }
    TreeNode* fun(vector<int>&post,vector<int>&in,int start,int end){
        if(start>end) return NULL;
        TreeNode* root=new TreeNode(post[i]);
        
        int idx=find(post[i],in);
        i--;
        if(idx==-1) return NULL;
        root->right=fun(post,in,idx+1,end);
        root->left=fun(post,in,start,idx-1);
       
        return root;

       
    }
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        
        int n=postorder.size();
        i=n-1;
        return fun(postorder,inorder,0,n-1);
        
    }
};

File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.i = 0

    def fun(self, start, end, ino, post):
        if start > end:
            return None

        node = TreeNode(post[self.i])

        pos = -1

        for x in range(start, end + 1):
            if ino[x] == post[self.i]:
                pos = x
                break

        if pos == -1:
            return None

        self.i -= 1

        node.right = self.fun(pos + 1, end, ino, post)

        node.left = self.fun(start, pos - 1, ino, post)

        return node

    def buildTree(self, inorder: list[int], postorder: list[int]) -> TreeNode | None:
        n = len(inorder)

        self.i = n - 1

        return self.fun(0, n - 1, inorder, postorder)